File: javascript/tests/delegation.test.js
/**
 * Tests for delegation attestation.
 */

import { test, describe } from 'node:test';
import assert from 'node:assert';
import { generateKeyPair } from '../src/crypto.js';
import { Capability } from '../src/capability.js';
import { DelegationRole } from '../src/types.js';
import {
    createAttestation,
    verifyAttestation,
    verifyChainDepth,
} from '../src/delegation.js';

describe('createAttestation and verifyAttestation', () => {
    test('create and verify roundtrip', () => {
        const kp = generateKeyPair();
        const caps = [new Capability('read:data'), new Capability('write:report')];

        const att = createAttestation(
            kp.privateKeyPem,
            'maker-2026-01',
            'maker.com',
            DelegationRole.MAKER,
            'urn:agentpin:maker.com:agent-type',
            'deployer.com',
            'urn:agentpin:deployer.com:instance',
            caps
        );

        assert.strictEqual(att.domain, 'maker.com');
        assert.strictEqual(att.role, DelegationRole.MAKER);

        assert.doesNotThrow(() => verifyAttestation(
            att, kp.publicKeyPem,
            'deployer.com', 'urn:agentpin:deployer.com:instance', caps
        ));
    });

    test('wrong key rejected', () => {
        const kp1 = generateKeyPair();
        const kp2 = generateKeyPair();
        const caps = [new Capability('read:data')];

        const att = createAttestation(
            kp1.privateKeyPem, 'kid', 'domain',
            DelegationRole.MAKER, 'agent',
            'delegatee', 'delegatee-agent', caps
        );

        assert.throws(
            () => verifyAttestation(att, kp2.publicKeyPem, 'delegatee', 'delegatee-agent', caps),
            /failed/i
        );
    });

    test('wrong capabilities rejected', () => {
        const kp = generateKeyPair();
        const caps = [new Capability('read:data')];
        const wrongCaps = [new Capability('write:data')];

        const att = createAttestation(
            kp.privateKeyPem, 'kid', 'domain',
            DelegationRole.MAKER, 'agent',
            'delegatee', 'delegatee-agent', caps
        );

        assert.throws(
            () => verifyAttestation(att, kp.publicKeyPem, 'delegatee', 'delegatee-agent', wrongCaps),
            /failed/i
        );
    });
});

describe('verifyChainDepth', () => {
    test('within depth passes', () => {
        assert.doesNotThrow(() => verifyChainDepth(1, [2, 3]));
        assert.doesNotThrow(() => verifyChainDepth(2, [2, 3]));
    });

    test('exceeding depth throws', () => {
        assert.throws(() => verifyChainDepth(3, [2, 3]), /exceeds/i);
    });
});

File: go/pkg/delegation/delegation_test.go
package delegation

import (
	"testing"

	"github.com/ThirdKeyAi/agentpin/go/pkg/crypto"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func TestCreateAndVerifyAttestation(t *testing.T) {
	kp, _ := crypto.GenerateKeyPair()
	priv, _ := crypto.LoadPrivateKey(kp.PrivateKeyPEM)
	pub, _ := crypto.LoadPublicKey(kp.PublicKeyPEM)
	caps := []types.Capability{"read:data", "write:report"}

	att, err := CreateAttestation(priv, "k", "maker.com", types.RoleMaker,
		"urn:maker:type", "deployer.com", "urn:deployer:inst", caps)
	if err != nil {
		t.Fatal(err)
	}
	if att.Domain != "maker.com" || att.Role != types.RoleMaker {
		t.Fatal("att fields wrong")
	}
	if err := VerifyAttestation(att, pub, "deployer.com", "urn:deployer:inst", caps); err != nil {
		t.Fatalf("verify failed: %v", err)
	}
}

func TestVerifyAttestationWrongKey(t *testing.T) {
	k1, _ := crypto.GenerateKeyPair()
	k2, _ := crypto.GenerateKeyPair()
	p1, _ := crypto.LoadPrivateKey(k1.PrivateKeyPEM)
	pub2, _ := crypto.LoadPublicKey(k2.PublicKeyPEM)
	caps := []types.Capability{"read:data"}
	att, _ := CreateAttestation(p1, "k", "d", types.RoleMaker, "a", "dd", "da", caps)
	if err := VerifyAttestation(att, pub2, "dd", "da", caps); err == nil {
		t.Fatal("wrong key should fail")
	}
}

func TestVerifyAttestationWrongCaps(t *testing.T) {
	kp, _ := crypto.GenerateKeyPair()
	priv, _ := crypto.LoadPrivateKey(kp.PrivateKeyPEM)
	pub, _ := crypto.LoadPublicKey(kp.PublicKeyPEM)
	caps := []types.Capability{"read:data"}
	bad := []types.Capability{"write:data"}
	att, _ := CreateAttestation(priv, "k", "d", types.RoleMaker, "a", "dd", "da", caps)
	if err := VerifyAttestation(att, pub, "dd", "da", bad); err == nil {
		t.Fatal("wrong caps should fail")
	}
}

func TestVerifyChainDepth(t *testing.T) {
	if err := VerifyChainDepth(1, []uint8{2, 3}); err != nil {
		t.Fatal(err)
	}
	if err := VerifyChainDepth(2, []uint8{2, 3}); err != nil {
		t.Fatal(err)
	}
	if err := VerifyChainDepth(3, []uint8{2, 3}); err == nil {
		t.Fatal("3 > min(2,3) should fail")
	}
}
